orangecontrib/bio/widgets3/OWdictyExpress.py

```diff
@@ -0,0 +1,235 @@
+                                      callback=self.auth_changed)
+
+        self.passfield.setEchoMode(QLineEdit.Password)
+
+        self.controlArea.layout().addWidget(h_line())
+
+        self.commit_button = gui.button(self.controlArea, self, "Commit", callback=self.commit)
+        self.handle_commit_button(False)
+
+        self.refresh_button = gui.button(self.controlArea, self, "Refresh", callback=self.refresh)
+        self.handle_cache_button(True)
+
+        gui.rubber(self.controlArea)
+
+        # Experiment Section
+
+        label = QLabel("Available projects:")
+        my_font = QFont()
+        my_font.setBold(True)
+        label.setFont(my_font)
+        self.mainArea.layout().addWidget(label)
+
+        self.mainArea.layout().addWidget(h_line())
+
+        self.filter = gui.lineEdit(self.mainArea, self, "searchString", "Filter:", callbackOnType=True,
+                                   callback=self.search_update)
+
+        self.experimentsWidget = QTreeWidget()
+        self.experimentsWidget.setAlternatingRowColors(True)
+        self.experimentsWidget.setRootIsDecorated(False)
+        self.experimentsWidget.setUniformRowHeights(True)
+        self.experimentsWidget.setSortingEnabled(True)
+
+        self.experimentsWidget.setItemDelegateForColumn(
+            0, gui.IndicatorItemDelegate(self, role=Qt.DisplayRole))
+
+        self.experimentsWidget.selectionModel().selectionChanged.connect(
+            self.onSelectionChanged)
+
+        self.experimentsWidget.setHeaderLabels(self.headerLabels)
+        self.mainArea.layout().addWidget(self.experimentsWidget)
+
+        if self.username and self.password:
+            self.connect()
+
+    def auth_set(self):
+        if len(self.username):
+            self.passfield.setDisabled(False)
+        else:
+            self.passfield.setDisabled(True)
+
+    def auth_changed(self):
+        self.auth_set()
+        self.connect()
+
+    def refresh(self):
+        self.reset()
+        self.load_experiments()
+
+    def reset(self):
+        self.experimentsWidget.clear()  # clear QTreeWidget
+        self.items = []
+        self.lastSelected = None
+        self.searchString = ""
+        self.handle_commit_button(False)
+
+    def search_update(self):
+        for item in self.items:
+            item.setHidden(
+                not all(s in item for s in self.searchString.split()))
+
+    def connect(self):
+        self.res = None
+        self.error(1)
+        self.reset()
+        self.handle_commit_button(False)
+        self.handle_cache_button(False)
+
+        if self.username and self.password:
+            try:
+                self.res = Genapi(self.username, self.password, self.server)
+            except Exception as e:
+                self.error(1, e.args[0])
+            else:
+                self.load_experiments()
+                self.handle_cache_button(True)
+
+    def load_experiments(self):
+        pbar = gui.ProgressBar(self, iterations=100)
```

Comment:
Progress bar is smart(ish). If you do not know the number of iterations, leave it empty and it will default to 100. If you know the number of iterations, you can set it and then just iterate until that number. It will automatically  report percentage. Read more about the progress bar here http://orange.readthedocs.io/en/latest/extend-widgets/rst/progressbar.html#. You can find more docs and examples in [the code](https://github.com/biolab/orange3/blob/9d08c66661f186f51e3df9ac2ca6277ed750b923/Orange/widgets/utils/progressbar.py).